### ISSUE TITLE
Make `round` and `signif` with an `Int` argument return an `Int`

### ISSUE THREAD

**Issue by GlenHertz** [JuliaLang/julia#23182]
Currently we have the following rounding routines for Int values:
```julia
julia> round(73489, 3)
73489.0
julia> round(73489, -3)
73000.0
julia> signif(73489, 3)
73500.0
```

There seems to be no reason to switch these to floating point.  I think it would be better to keep them as Int and get the following behavior:

```julia
julia> round(73489, 3)  # already perfectly rounded
73489
julia> round(73489, -3)
73000
julia> signif(73489, 3)
73500
```
Also, if people want to do Int rounding we could also have:

```julia
julia> round(Int, 73489.456, 1) 
73490
julia> round(Int, 73489.456, -3)
73000
julia> signif(Int, 73489, 3)
73500
```

Does this API seem reasonable?

**Comment by simonbyrne** [comment]
`round(Int, 73489.456, 1)` should probably return an error.

**Comment by GlenHertz** [comment]
Yeah, it is a bit of a contradiction so maybe good to raise an error.  The following seems a bit silly:

```julia
julia> round(Int, 73489.456, 1)
73490
julia> round(Int, 73489.456, 0)
73489
```

I don't see a good use case for `round(Int, 343.343, $pos_integer)` unless maybe if someone is trying to write generic code.

**Comment by simonbyrne** [comment]
The main challenge is figuring out a way to do it in a generic fashion, avoiding intermediate float rounding and integer overflow.

**Comment by GlenHertz** [comment]
An issue with the current interface is the second argument is specific to rounding to decimals which is not generic.  I think a better if the second argument was the step size of rounding (and the resultant type).  For example:

```julia
julia> round(10.27, 0.1)
10.3
julia> round(10.27, 0.25)
10.25
julia> round(10.27, 1)
10
julia> round(11//16, 1//4)
3//4
julia> round(1234, 100)
1200
```

It should take on the type of the second arg since the number is rounding to its resolution.

I think the `eltype` form is not needed as the eltype of the result is the same as the eltype of the second argument.

```julia
julia> round(9.45, 1)
9
julia> round(9.45, 1.0)
9.0
julia> round(9.45, 1//2)
19//2
```

Probably too big of a change to make.  Any issues with this?

**Comment by nalimilan** [comment]
> An issue with the current interface is the second argument is specific to rounding to decimals which is not generic.

That's not really an issue since rounding an integer will give an integer whatever the number of decimal places requested. So that should always work. OTOH it's fine to have ` round(Int, 343.343, d)` fail.

Rounding to a given number of decimal places is a very standard operation, I'd rather not use a different approach that people are not familiar with.

**Comment by adienes** [comment]
every single function call in this issue is now an error 😀

**Comment by inkydragon** [comment]
- `signif` has been deprecated in favor of the `sigdigits` keyword argument to round.
 in Julia v0.7
	https://github.com/JuliaLang/julia/blob/master/HISTORY.md#julia-v070-release-notes

```jl
julia> round(73489, 3)
ERROR: MethodError: no method matching round(::Int64, ::Int64)

julia> round(73489, -3)
ERROR: MethodError: no method matching round(::Int64, ::Int64)

julia> round(73489; sigdigits=3)
73500.0
```

```jl
julia> round(10.27, 0.1)
ERROR: MethodError: no method matching round(::Float64, ::Float64)

julia> round(10.27, 1)
ERROR: MethodError: no method matching round(::Float64, ::Int64)

julia> round(11//16, 1//4)
ERROR: MethodError: no method matching round(::Rational{Int64}, ::Rational{Int64})

julia> round(1234, 100)
ERROR: MethodError: no method matching round(::Int64, ::Int64)
```

```jl
julia> round(Int, 73489.456, RoundUp)
73490

julia> round(Int, 73489.456, RoundDown)
73489
```
